Add ecobee fan mode

uchagani opened the pull request:
## Description:  

* Adds the ability to turn the fan `on` on ecobee thermostats.
* Updated the temperature logic with best practice from ecobee api.
* added a new `fan_list` to the attributes similar to the base class
* when in `non-auto` mode, the `target_temp_high` and `target_temp_low` attributes will no longer be `null`.  Instead they will reflect the `temperature` attribute.

## Example entry for `configuration.yaml` (if applicable):
```yaml
ecobee:
  api_key: super_secret_api_key
```

## Checklist:
  - [X] The code change is tested and works locally.

If the code communicates with devices, web services, or third-party tools:
  - [X] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [X] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [X] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [X] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.

If the code does not interact with devices:
  - [X] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [X] Tests have been added to verify that the new code works.

Commit:
add ability to set fan on

Commit:
add tests and change "not on" status to "auto"

Commit:
hound fix

Commit:
more hounds

Commit:
I don't understand new lines

Commit:
fix linting errors

Commit:
more linting fixes

Commit:
change method signature

Commit:
lint fixes

Commit:
hopefully last lint fix

Commit:
correct temp ranges according to ecobee API docs

Commit:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into add-ecobee-fan-mode

Commit:
update dependency to latest version

uchagani commented:
@technicalpickles New PR with updated pyecobee code and other changes we discussed.

Commit:
update tests with values from new temp logic

Commit:
fix linting issue

Commit:
more linting fixes

andrey-git commented:
Should the fan be reflected in supported bitmap?

Commit:
add SUPPORT_FAN_MODE to capabilities

uchagani commented:
good catch @andrey-git.  I added the `SUPPORT_FAN_MODE` flag

technicalpickles commented on homeassistant/components/climate/ecobee.py:
```diff
@@ -190,7 +190,7 @@ def fan(self):
         """Return the current fan state."""
         if 'fan' in self.thermostat['equipmentStatus']:
             return STATE_ON
-        return STATE_OFF
```
Is equipmentStatus what is _currently_ running, rather than what is configured? If that's the case, then I think it should stay `STATE_OFF`.

uchagani replied:
@technicalpickles you're correct, this is the currently running status so I'll change this back to `STATE_OFF`, however it looks like the method name for this is incorrect and that's why we see `null` for `fan_mode` in the attributes.  I'm fixing that too.  should be pushed soon.

uchagani commented:
@technicalpickles need your opinion on something.

Right now, if the thermostat is on anything but `auto`, then `target_temp_high` and `target_temp_low` will be `null` in the attributes.  I'm considering populating these with the values that come back from the thermostat.  Which means it would be set to the same value as the `temperature`.  This follows Ecobee's best practice of setting high/low to be the same when not in auto mode.  Thoughts?

Commit:
add fan_list to attributes.
restore current fan state to OFF if fan is not running.
change target high/low temps from null to target temp when not in auto mode.
change target temp from null to high/low temp when in auto mode
change mode attribute to climate_mode for consistency with other lists.

Commit:
remove unused import

Commit:
simplify logic

uchagani commented:
@technicalpickles I've made the changes.

* `fan` now returns the current state of the fan (either `on` or `off`).  
* added a new `fan_list` to the attributes similar to the base class
* when in `non auto` mode, the `target_temp_high` and `target_temp_low` attributes will no longer be null.  instead they will reflect the `temperature` attribute.

Commit:
lint fixes

technicalpickles commented:
@uchagani that makes sense to me 👍  I would check how this looks in the UI, in case there is different behavior for when there is a target high/low AND target_temperature, and also check how other climates do this.

uchagani commented:
@technicalpickles the UI is kinda messed up right now but I confirmed with someone else and it has to do with HA overall.  Here's what the latest version of HA looks like: 
![image](https://user-images.githubusercontent.com/1053584/36765110-ad989fbc-1bfd-11e8-9700-e4d142761e50.png)

Again, this is just a bug on the UI itself and not related to this PR.

technicalpickles commented:
What happens if you adjust target temp (without high/low) while in auto? I think that is why those were `None` before, so the UI doesn't show them. Either way, probably good to file an issue about that separate from this 😅

uchagani commented:
Here's how the UI looks right now if the thermostat is in `non-auto` mode (heat/cool/etc) and the `target_temp_high` and `target_temp_low` are null (as before).  Same UI bug.

![image](https://user-images.githubusercontent.com/1053584/36767649-69744c26-1c09-11e8-8de7-3cb50a9804dc.png)

uchagani commented:
@technicalpickles the issue has already been created and there are some architecture discussions happening on it here: https://github.com/home-assistant/architecture/issues/1

technicalpickles commented:
I've been using branch (plus latest dev merged) on my 'production' homeassistant branch to do some automation, and it seems to be working well. I mentioned the main caveat over in https://github.com/home-assistant/home-assistant/pull/12606:

---

I am seeing one other oddity though, when using `climate.set_fan_mode`. Consider this script:

```
script:
  climate_its_ok_out:
    sequence:
      - service: climate.set_operation_mode
        entity_id: climate.home
        data:
          operation_mode: 'auto'
      - service: climate.ecobee_resume_program
        entity_id: climate.home
      - service: climate.set_fan_mode
        entity_id: climate.home
        data:
          fan_mode: auto
```

When this is called, it correct sets operation mode to auto, and resumes the program (although now I'm looking it, I probably only need the latter). However, the temperature remains what it is. If I call `climate.set_fan_mode` first, `climate.ecobee_resume_program`, it works as expected.

I think this has to do with a detail of ecobee's set_fan_mode, where their API doesn't distinguish between setting the temperature and setting the fan mode. That is, setting the fan mode requires you specify the temperature also. At the time that `climate.set_fan_mode`, it doesn't seem that the local ecobee data is refreshed at that point, so the temperature gets set as temperature homeassistant thinks it is, rather than what the ecobee remote API thinks it is.

There are a least a  few possible workarounds as a user. The first I mentioned, making sure to `climate.set_fan_mode` before changing the temperature to ensure the temperature is called later. Another _might_ be to use a `delay` to try to wait until the local state has updated, but that is probably a race condition.

I'm not sure what else we can do on the backend. It might make sense to remove the `ecobee.set_fan_mode` method, and instead make `set_temperature` method in this branch take a keyword arg for the fan mode, and it passes it down to all of the the way to the python-ecobee-api call (set_temp_hold or set_auto_hold). That would involve more changes to this branch and python-ecobee-api and this branch though 😞 

---

Do you think that would be a blocker to getting this merged?

uchagani commented:
I think we can merge this in as is. The delay of the ecobee API is just something we will have to live with because even if you update from your ecobee app, it takes a bit for the thermostat to update.

uchagani commented:
@andrey-git @balloob Does this PR need anything else before it can be merged?

balloob commented on homeassistant/components/climate/ecobee.py:
```diff
@@ -158,16 +159,12 @@ def current_temperature(self):
-        if self.current_operation == STATE_AUTO:
-            return self.thermostat['runtime']['desiredHeat'] / 10.0
-        return None
+        return self.thermostat['runtime']['desiredHeat'] / 10.0
```
We shouldn't do this, although I guess we should filter this out at the climate level. There is no upper/lower bound when we just have a target temp.

balloob commented:
ok to merge once target temp low/high has been reverted.

Commit:
revert change for target temps

uchagani commented:
@balloob target temp low/high has been reverted.

roch23 commented:
Appreciate this! Had previously been controlling with set_fan_min_on_time, which frequently would fail based on the operating mode. As an ugly workaround I would store the current temp, set the fan, then restore the temp as a temporary hold - it worked but caused issues with away mode / etc.
